core/src/main/java/io/grpc/Context.java

```diff
@@ -0,0 +1,608 @@
+        public void close(ScheduledExecutorService instance) {
+          instance.shutdown();
+        }
+
+        @Override
+        public String toString() {
+          return name;
+        }
+      };
+
+  /**
+   * Stack of context objects which is used to record attach & detach history on a thread.
+   */
+  private static final ThreadLocal<ArrayDeque<Context>> contextStack =
+      new ThreadLocal<ArrayDeque<Context>>() {
+    @Override
+    protected ArrayDeque<Context> initialValue() {
+      return new ArrayDeque<Context>();
+    }
+  };
+
+  private static final Object[][] EMPTY_ENTRIES = new Object[0][2];
+
+  /**
+   * The logical root context which is {@link #current()} if no other context is bound. This context
+   * is not cancellable and so will not cascade cancellation or retain listeners.
+   */
+  public static final Context ROOT = new Context(null);
+
+  /**
+   * Create a {@link Key} with the given name.
+   */
+  public static <T> Key<T> key(String name) {
+    return new Key<T>(name);
+  }
+
+  /**
+   * Create a {@link Key} with the given name and default value.
+   */
+  public static <T> Key<T> keyWithDefault(String name, T defaultValue) {
+    return new Key<T>(name, defaultValue);
+  }
+
+  /**
+   * Return the context associated with the current thread, will never return {@code null} as
+   * the {@link #ROOT} context is implicitly associated with all threads.
+   *
+   * <p>Will never return {@link CancellableContext} even if one is attached, instead a
+   * {@link Context} is returned with the same properties and lifetime. This is to avoid
+   * code stealing the ability to cancel arbitrarily.
+   */
+  public static Context current() {
+    ArrayDeque<Context> stack = contextStack.get();
+    if (stack.isEmpty()) {
+      return ROOT;
+    }
+    return stack.peekLast();
+  }
+
+  private final Context parent;
+  private final Object[][] keyValueEntries;
+  private final ArrayList<ExecutableListener> listeners;
+  private final boolean cascadesCancellation;
+
+  /**
+   * Construct a context that cannot be cancelled and will not cascade cancellation from its parent.
+   */
+  private Context(Context parent) {
+    this.parent = parent;
+    keyValueEntries = EMPTY_ENTRIES;
+    listeners = null;
+    cascadesCancellation = false;
+  }
+
+  /**
+   * Construct a context that cannot be cancelled but will cascade cancellation from its parent if
+   * it is cancellable.
+   */
+  private Context(Context parent, Object[][] keyValueEntries) {
+    this.parent = parent;
+    this.keyValueEntries = keyValueEntries;
+    cascadesCancellation = true;
+    listeners = new ArrayList<ExecutableListener>();
+    if (parent.canBeCancelled()) {
+      parent.addListener(new CancellationListener() {
+        @Override
+        public void cancelled(Context context) {
+          notifyAndClearListeners();
+        }
+      }, MoreExecutors.directExecutor());
+    }
+  }
+
+  protected boolean canBeCancelled() {
```

Comment:
Using package-private instead of protected would mean that these methods wouldn't generally show up in JavaDoc, which would seem beneficial.